src/openrct2/world/Banner.cpp

```diff
@@ -255,60 +285,83 @@ void BannerResetBrokenIndex()
 {
-    // For each banner in the map, check if the banner index is in use already, and if so, create a new entry for it
-    std::vector<bool> activeBanners;
-    activeBanners.resize(MAX_BANNERS);
+    // Sort the banners by index
+    std::sort(bannerElements.begin(), bannerElements.end(), [](const BannerElementWithPos& a, const BannerElementWithPos& b) {
+        return a.Element->GetIndex() < b.Element->GetIndex();
+    });
+
+    // Create a list of all banners with duplicate indices.
+    std::vector<BannerElementWithPos> duplicates;
+    for (size_t i = 1; i < bannerElements.size(); i++)
+    {
+        if (bannerElements[i - 1].Element->GetIndex() == bannerElements[i].Element->GetIndex())
+        {
+            duplicates.push_back(bannerElements[i]);
+        }
+    }
 
-    for (int y = 0; y < gMapSize.y; y++)
+    // For each duplicate, create a new banner and copy the old data
+    for (const auto& duplicate : duplicates)
     {
-        for (int x = 0; x < gMapSize.x; x++)
+        const auto oldIndex = duplicate.Element->GetIndex();
+        const auto* oldBanner = GetBanner(oldIndex);
+        if (oldBanner == nullptr)
         {
-            const auto bannerPos = TileCoordsXY{ x, y }.ToCoordsXY();
-            for (auto* bannerElement : OpenRCT2::TileElementsView<BannerElement>(bannerPos))
-            {
-                auto bannerIndex = bannerElement->GetIndex();
-                if (bannerIndex == BannerIndex::GetNull())
-                    continue;
+            LOG_ERROR("Unable to get old banner for index %u.", oldIndex.ToUnderlying());
+            continue;
+        }
 
-                const auto index = bannerIndex.ToUnderlying();
-                if (activeBanners[index])
-                {
-                    LOG_INFO(
-                        "Duplicated banner with index %d found at x = %d, y = %d and z = %d.", index, x, y,
-                        bannerElement->BaseHeight);
-
-                    // Banner index is already in use by another banner, so duplicate it
-                    auto newBanner = CreateBanner();
-                    if (newBanner == nullptr)
-                    {
-                        LOG_ERROR("Failed to create new banner.");
-                        continue;
-                    }
-                    Guard::Assert(!activeBanners[newBanner->id.ToUnderlying()]);
-
-                    // Copy over the original banner, but update the location
-                    const auto* oldBanner = GetBanner(bannerIndex);
-                    if (oldBanner != nullptr)
-                    {
-                        auto newBannerId = newBanner->id;
-
-                        *newBanner = *oldBanner;
-                        newBanner->id = newBannerId;
-                        newBanner->position = { x, y };
-                    }
-
-                    bannerElement->SetIndex(newBanner->id);
-                }
+        auto newBanner = CreateBanner();
```

Comment:
Put const auto*

Reply:
I need to write into it, so it can't be const.